resources/type-system/inference.md

```diff
@@ -1117,6 +1134,26 @@ succintly, the syntax of Dart is extended to allow the following forms:
   any loss of functionality, provided they are not trying to construct a proof
   of soundness._
 
+In addition, the following forms are added to allow constructor invocations,
+dynamic method invocations, function object invocations, instance method
+invocations, and static/toplevel method invocations to be distinguished:
+
+- `@DYNAMIC_INVOKE(m_0.id<T_1, T_2, ...>(n_1: m_1, n_2: m_2, ...))`
+
+- `@FUNCTION_INVOKE(m_0.call<T_1, T_2, ...>(n_1: m_1, n_2: m_2, ...))`
+
+- `@INSTANCE_INVOKE(m_0.id<T_1, T_2, ...>(n_1: m_1, n_2: m_2, ...))`
+
+- `@STATIC_INVOKE(f<T_1, T_2, ...>(n_1: m_1, n_2: m_2, ...))`
+
+In each of these forms, each `m_i` represents an elaborated expression, each
+`T_i` represents a type, and each `n_i` represents an optional argument name
+identifier. When present, `id` represents an identifier or operator name, and
+`f` represents a static method or top level function.
+
+The semantics of each of these forms is to evaluate the `m_i` in sequence, then
```

Comment:
Technically I think "evaluating the lookup of `id` in `m_i`" is interposed in here.  This is required for at least the DYNAMIC_INVOKE, since you don't know whether you are invoking a method or calling a function typed getter.  Whether it is required for any of the others depends a bit on the answer to my previous question about what semantic forms are intended to be covered here.  That is, do you intend 
```
class A {
  int Function(int) f = throw "yeehaw"; 
  void test() {
      this.f(3);
  }
}
```
to be covered by one of the above, or do you intend to add something else to cover this later?

Reply:
Ok, I’ve tried to clarify this as much as I can without formally specifying the exact semantics of each form.

As to your example with `this.f(3)`, I intend to eventually represent this as something like `@FUNCTION_INVOKE(@INSTANCE_GET(m_0.id).call(3))`. In this section, I’ve tried to hint at this in the non-normative text I’ve added under `@FUNCTION_INVOKE`.

The full details are in the “Method invocation inference” section, under the bullet “Otherwise, if `U_0` has an accessible instance getter named `id`, then:”, but that text is a little wishy washy because I haven’t introduced the notion of `@INSTANCE_GET` yet as of this PR.